frontend/src/components/management/WorkflowAssignment.tsx

```diff
@@ -0,0 +1,454 @@
+    interface ActionInfo {
+        study: Pick<FMSStudy, 'id' | 'letter'>
+        workflow: FMSWorkflow['name']
+        step: Step['name']
+        stepOrder: FMSSampleNextStepByStudy['step_order'],
+        alreadyQueued: Sample['id'][]
+    }
+    const [dequeueActions, setDequeueActions] = useState<ActionInfo[]>([])
+    const [queueActions, setQueueActions] = useState<ActionInfo[]>([])
+
+    interface StudyWorkflow {
+        study: Study['letter']
+        workflow: Workflow['name']
+    }
+    const [studyWorkflowsByProject, setStudyWorkflowsByProject] = useState<Record<Project['id'], StudyWorkflow[]>>({})
+
+    interface CommonProject {
+        id: Project['id']
+        name: Project['name']
+    }
+    const [commonProjects, setCommonProjects] = useState<CommonProject[]>([])
+
+    const [selectedProject, setSelectedProject] = useState<Project['id']>()
+    const [selectedStudyWorkflow, setSelectedStudyWorkflow] = useState<StudyWorkflow>()
+
+    const refreshActions = useCallback(async () => {
+        const samples = await dispatch(fetchSamplesByDefaultSelectionAndExceptedIDs(defaultSelection, exceptedSampleIDs, filters))
+        const sampleIDs = samples.map(sample => sample.id)
+
+        if (samples.length === 0) {
+            setQueueActions([])
+            setDequeueActions([])
+            setStudyWorkflowsByProject({})
+            setCommonProjects([])
+            setIsFetching(false)
+            setSelectedProject(undefined)
+            setSelectedStudyWorkflow(undefined)
+            return
+        }
+
+        setIsFetching(true)
+
+        const pooledSamples = (await dispatch(api.pooledSamples.list({ sample__id__in: sampleIDs.join(',') }))).data.results
+        const projectsBySample = [...samples, ...pooledSamples].reduce<Record<Sample['id'], Set<Project['id']>>>((acc, sample) => {
+            if ('pool_id' in sample) {
+                // handle pooled samples
+                if (!acc[sample.pool_id]) {
+                    acc[sample.pool_id] = new Set()
+                }
+                acc[sample.pool_id].add(sample.project_id)
+            } else {
+                // handle non-pooled samples
+                if (!acc[sample.id]) {
+                    acc[sample.id] = new Set()
+                }
+                if (sample.project) {
+                    acc[sample.id].add(sample.project)
+                }
+            }
+            return acc
+        }, {})
+
+        type StudyStepKey = `${Study['id']}-${Step['name']}`
+        const sampleNextSteps = (await dispatch(api.sampleNextStep.listSamples(sampleIDs))).data.results
+        const studyStepCount: Record<string, number> = {}
+        const sampleIDsByStudyStep: Record<StudyStepKey, Set<Sample['id']>> = {}
+        for (const sampleNextStep of sampleNextSteps) {
+            for (const study of sampleNextStep.studies) {
+                const key: StudyStepKey = `${study}-${sampleNextStep.step.name}`
```

Comment:
study step should not be identified by step name ... step name can be present multiple time in a workflow study. I don't see a logic where this would not be important ...You are likely to prevent queueing samples to steps they are not currently queued to.